action/smart_contract/execute.go

```diff
@@ -0,0 +1,166 @@
+package smart_contract
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Oneledger/protocol/action"
+	"github.com/Oneledger/protocol/action/helpers"
+	"github.com/Oneledger/protocol/data/keys"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/pkg/errors"
+	"github.com/tendermint/tendermint/libs/kv"
+)
+
+type Execute struct {
+	From   action.Address  `json:"from"`
+	To     *action.Address `json:"to"`
+	Amount action.Amount   `json:"amount"`
+	Data   []byte          `json:"data"`
+	Nonce  uint64          `json:"nonce"`
+}
+
+func (e Execute) Marshal() ([]byte, error) {
+	return json.Marshal(e)
+}
+
+func (e *Execute) Unmarshal(data []byte) error {
+	return json.Unmarshal(data, e)
+}
+
+func (e Execute) Signers() []action.Address {
+	return []action.Address{e.From.Bytes()}
+}
+
+func (e Execute) Type() action.Type {
+	return action.SC_EXECUTE
+}
+
+func (e Execute) Tags() kv.Pairs {
+	tags := make([]kv.Pair, 0)
+
+	tag := kv.Pair{
+		Key:   []byte("tx.type"),
+		Value: []byte(e.Type().String()),
+	}
+	tag2 := kv.Pair{
+		Key:   []byte("tx.owner"),
+		Value: e.From.Bytes(),
+	}
+	tags = append(tags, tag, tag2)
+	if e.To != nil {
+		tags = append(tags, kv.Pair{
+			Key:   []byte("tx.to"),
+			Value: e.To.Bytes(),
+		})
+	}
+	return tags
+}
+
+var _ action.Tx = scExecuteTx{}
+
+type scExecuteTx struct {
+}
+
+func (s scExecuteTx) Validate(ctx *action.Context, tx action.SignedTx) (bool, error) {
+	execute := &Execute{}
+	err := execute.Unmarshal(tx.Data)
+	if err != nil {
+		return false, errors.Wrap(action.ErrWrongTxType, err.Error())
+	}
+	//validate basic signature
+	err = action.ValidateBasic(tx.RawBytes(), execute.Signers(), tx.Signatures)
+	if err != nil {
+		return false, err
+	}
+
+	err = action.ValidateFee(ctx.FeePool.GetOpt(), tx.Fee)
+	if err != nil {
+		return false, err
+	}
+
+	//validate transaction specific field
+	if !execute.Amount.IsValid(ctx.Currencies) {
+		return false, errors.Wrap(action.ErrInvalidAmount, execute.Amount.String())
+	}
+
+	if execute.From.Err() != nil || execute.To != nil && execute.To.Err() != nil {
+		return false, action.ErrInvalidAddress
+	}
+
+	if len(execute.Data) == 0 {
+		return false, action.ErrMissingData
+	}
+	return true, nil
+}
+
+func (s scExecuteTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	ctx.Logger.Detail("Processing SC Deploy Transaction for CheckTx", tx)
+	// NOTE: We do not need to run a logic, but only calculate a gas
+	// maybe it will be required a stateDB copy without deliver and clear it's dirties
+	// TODO: Add gas calculation here
+	return true, action.Response{GasUsed: -1}
```

Comment:
why not directly run the tx, the checkstate will be dropped as you expected.

Reply:
Because it will run a evm code which is not a part of the txChecker. Also it breaks a state for the stateDB and we have an API to calculate the aprox gas.

Reply:
Also by the tendermint what I found that we do not need a full logic during check

Reply:
for tendermint, it doesn't require checktx to execute full logic, but as a blockchain, the checktx result different from delivery tx might cause apphash difference in certain cases.

Reply:
@alexjlan https://docs.tendermint.com/master/spec/abci/abci.html#checktx
```
CheckTx need not execute the transaction in full, but rather a light-weight yet stateful validation, like checking signatures and account balances, but not running code in a virtual machine.
```
From the code above it is deterministic, but in the case of VM the statedb needs to be encapsulated as the bunch of logic are executed at the end (blockEnder). If it will be running there, it may cause the apphash difference as:
```
The transaction may come from an external user or another node
```
as some state is not already on the node, which will execute it.

I can add more checks related to the state, but rather than run a full evm execution. Also it will slow down the network in case of high gas limit contracts.

Reply:
Added